go/vt/tabletserver/engines/schema/schema_engine.go

```diff
@@ -0,0 +1,489 @@
+	se.mu.Lock()
+	defer se.mu.Unlock()
+	if !se.isOpen {
+		return nil
+	}
+	defer tabletenv.LogError()
+
+	curTime, tableData, err := func() (int64, *sqltypes.Result, error) {
+		conn, err := se.conns.Get(ctx)
+		if err != nil {
+			return 0, nil, tabletenv.NewTabletErrorSQL(vtrpcpb.ErrorCode_INTERNAL_ERROR, err)
+		}
+		defer conn.Recycle()
+		curTime, err := se.mysqlTime(ctx, conn)
+		if err != nil {
+			return 0, nil, err
+		}
+		tableData, err := conn.Exec(ctx, mysqlconn.BaseShowTables, maxTableCount, false)
+		if err != nil {
+			return 0, nil, err
+		}
+		return curTime, tableData, nil
+	}()
+	if err != nil {
+		return fmt.Errorf("could not get table list for reload: %v", err)
+	}
+
+	// Reload any tables that have changed. We try every table even if some fail,
+	// but we return success only if all tables succeed.
+	// The following section requires us to hold mu.
+	rec := concurrency.AllErrorRecorder{}
+	curTables := map[string]bool{"dual": true}
+	for _, row := range tableData.Rows {
+		tableName := row[0].String()
+		curTables[tableName] = true
+		createTime, _ := row[2].ParseInt64()
+		// Check if we know about the table or it has been recreated.
+		if _, ok := se.tables[tableName]; !ok || createTime >= se.lastChange {
+			func() {
+				// Unlock so CreateOrAlterTable can lock.
+				se.mu.Unlock()
+				defer se.mu.Lock()
+				log.Infof("Reloading schema for table: %s", tableName)
+				rec.RecordError(se.CreateOrAlterTable(ctx, tableName))
+			}()
+			// In case someone closed se when lock was released.
+			if !se.isOpen {
+				return nil
+			}
+			continue
+		}
+		// Only update table_rows, data_length, index_length, max_data_length
+		se.tables[tableName].SetMysqlStats(row[4], row[5], row[6], row[7], row[8])
+	}
+	se.lastChange = curTime
+
+	// Handle table drops
+	var dropped []string
+	for tableName := range se.tables {
+		if curTables[tableName] {
+			continue
+		}
+		delete(se.tables, tableName)
+		dropped = append(dropped, tableName)
+	}
+	// We only need to broadcast dropped tables because
+	// CreateOrAlterTable will broadcast the other changes.
+	if len(dropped) > 0 {
+		se.broadcast(nil, nil, dropped)
+	}
+	return rec.Error()
+}
+
+func (se *SchemaEngine) mysqlTime(ctx context.Context, conn *connpool.DBConn) (int64, error) {
+	tm, err := conn.Exec(ctx, "select unix_timestamp()", 1, false)
+	if err != nil {
+		return 0, tabletenv.PrefixTabletError(vtrpcpb.ErrorCode_UNKNOWN_ERROR, err, "Could not get MySQL time: ")
+	}
+	if len(tm.Rows) != 1 || len(tm.Rows[0]) != 1 || tm.Rows[0][0].IsNull() {
+		return 0, tabletenv.NewTabletError(vtrpcpb.ErrorCode_UNKNOWN_ERROR, "Unexpected result for MySQL time: %+v", tm.Rows)
+	}
+	t, err := strconv.ParseInt(tm.Rows[0][0].String(), 10, 64)
+	if err != nil {
+		return 0, tabletenv.NewTabletError(vtrpcpb.ErrorCode_UNKNOWN_ERROR, "Could not parse time %+v: %v", tm, err)
+	}
+	return t, nil
+}
+
+// CreateOrAlterTable must be called if a DDL was applied to that table.
```

Comment:
Maybe this should be changed to TableWasCreatedOrAltered and the next one to TableWasDropped?